components/director/internal/domain/formationtemplate/resolver.go

```diff
@@ -280,3 +298,47 @@ func (r *Resolver) Webhooks(ctx context.Context, obj *graphql.FormationTemplate)
 
 	return gqlWebhooks, nil
 }
+
+// FormationConstraint retrieves a FormationConstraint for the specified FormationTemplate
+func (r *Resolver) FormationConstraint(ctx context.Context, obj *graphql.FormationTemplate) ([]*graphql.FormationConstraint, error) {
+	params := dataloader.ParamFormationConstraint{ID: obj.ID, Ctx: ctx}
+	return dataloader.FormationTemplateFor(ctx).FormationConstraintByID.Load(params)
+}
+
+// FormationConstraintsDataLoader retrieves Formation Constraints for each FormationTemplate ID in the keys
+func (r *Resolver) FormationConstraintsDataLoader(keys []dataloader.ParamFormationConstraint) ([][]*graphql.FormationConstraint, []error) {
+	if len(keys) == 0 {
+		return nil, []error{apperrors.NewInternalError("No Formation Templates found")}
+	}
+
+	ctx := keys[0].Ctx
+	formationTemplateIDs := make([]string, 0, len(keys))
+	for _, key := range keys {
+		formationTemplateIDs = append(formationTemplateIDs, key.ID)
+	}
+
+	tx, err := r.transact.Begin()
+	if err != nil {
+		return nil, []error{err}
+	}
+	defer r.transact.RollbackUnlessCommitted(ctx, tx)
+
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	formationConstraintsPerFormation, err := r.formationConstraintSvc.ListByFormationTemplateIDs(ctx, formationTemplateIDs)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	gqlFormationConstraints := make([][]*graphql.FormationConstraint, 0, len(formationConstraintsPerFormation))
+
+	for _, formationConstraints := range formationConstraintsPerFormation {
```

Comment:
Additionally to the comment in the `formationconstraint/service.go`, this will be simplified as well and most probably will have a similar look:
```
func (r *Resolver) FormationConstraintsDataLoader(keys []dataloader.ParamFormationConstraint) ([]*graphql.FormationConstraint, []error) {
	if len(keys) == 0 {
		return nil, []error{apperrors.NewInternalError("No Formation Templates found")}
	}

	ctx := keys[0].Ctx
	formationTemplateIDs := make([]string, 0, len(keys))
	for _, key := range keys {
		formationTemplateIDs = append(formationTemplateIDs, key.ID)
	}

	tx, err := r.transact.Begin()
	if err != nil {
		return nil, []error{err}
	}
	defer r.transact.RollbackUnlessCommitted(ctx, tx)

	ctx = persistence.SaveToContext(ctx, tx)

	formationConstraints, err := r.formationConstraintSvc.ListByFormationTemplateIDs(ctx, formationTemplateIDs)
	if err != nil {
		return nil, []error{err}
	}

	gqlFormationConstraints := r.formationConstraintConverter.MultipleToGraphQL(formationConstraints)

	if err = tx.Commit(); err != nil {
		return nil, []error{err}
	}

	return gqlFormationConstraints, nil
}
```

Also, that way in the root_resolver we'll have `func (r *RootResolver) FormationConstraintsDataLoader(ids []dataloader.ParamFormationConstraint) ([]*graphql.FormationConstraint, []error) {}` which is similar to the other data loaders that we have, instead of `([][]*graphql.FormationConstraint, []error)`

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/31803034/229488239-08122b5e-1331-4909-a0bd-ed7bffca1474.png">


I may be missing something but at the first glance, I'm not sure we need this matrix structure in the service and resolver files?